### DISCUSSION TITLE
Dissolving item stacks in parallel or serial

### DISCUSSION THREAD

**Discussion by FlameArrow57** [stonepillars/goon-flock#210] [Ideas]
Discussion for whether item stacks should be broken down in parallel or serial

Items have a health variable that determines their health. A stack of items is a single item with an amount variable, that has the health of a single item.

Item stacks being broken down in parallel would mean that if dissolving a stack of items for resources, dissolving the whole stack of items is based on decrementing total item health, scaled by the number of items. It would make dealing with fractional health of items simpler. The animation for dissolving makes a bit more sense too. The issue is that it isn't super helpful for player feedback, which can be remedied by adding an item health hud bar next to the reclaimer which shows the item's total health.

Item stacks being broken down in serial would mean that if dissolving a stack of items for resources, dissolving is based both on item health and decrementing current amount in the stack. When the health of the item reaches zero, the health is reset to its initial health and the amount is decremented. It makes it a bit more tricky to deal with fractional health items as item health is determined after compile time. The animation for dissolving doesn't make as much sense here since it seems the whole item is being dissolved when only one in the stack is, but this could be remedied by only applying the animation for when there's one item. This would be better for visual feedback if the item health hud bar isn't used.

**Comment by stonepillars** [comment]
The problem with parallel eating is even if most of the items are eaten, if a drone drops it, it can still be used by other players normally since the amount remains unchanged. The problem with serial eating is you can't handle fractional health very well.
Maybe some sort of mix of both? Flame discussed storing the amount or health when the item is initially put into the reclaimer and then using that for the math and removing it from the items that way.

**Reply by FlameArrow57** [reply]
I think burning and acid dissolved stack item can be used normally too?

**Reply by FlameArrow57** [reply]
Was actually an issue with that for serial, was something amy said, "the problem with that is that if a flockdrone has partially eaten an item of the stack, then drops the stack, another drone would pick it up and it'd act like the whole stack was damaged rather than just the one item."

**Reply by stonepillars** [reply]
What if you un-stack a stack? Does the health carry over? or can you eat the same thing forever?

**Comment by amylizzle** [comment]
When you say fractional health, can you give an example of an item stack that has/would have it? Because if we're talking about an item that starts with 100 health, and it somehow gets 43.2 health, I think it's absolutely fine to just ignore that .2. In the scheme of things, it basically only results in a single extra tick of item dissolving, which is like 2 extra resources.

**Reply by FlameArrow57** [reply]
What I was thinking from burning code: Item health is 10, burning is 10 / 4 = 2.5, then next is 0.5 health.

**Reply by amylizzle** [reply]
I think that doesn't matter though? An extra .5 health is basically only worth an extra tick of the resource eating code. A player isn't going to notice, and it's not going  to affect anything in any tangible way except that *very* occasionally an item might give a couple more resource points than it ought to for its level of damage.

**Reply by FlameArrow57** [reply]
Fractional health also comes up if doing it in parallel, health of one item in the stack is I.health / I.amount.

**Reply by amylizzle** [reply]
I am really not understanding why this is an issue. Does it matter if we treat health as a float or an int? we're eating it a 2.0 per tick either way. It doesn't get displayed to the user (at least, in raw number form). Frankly the only impact would be a rounding error which might give one or two resources more/less than it should, but who's gonna notice?

Can you provide an example of when fractional health and talk me through why? I may just be missing the point, it's been a long week and I am tired.

**Reply by FlameArrow57** [reply]
It's not too much of an issue, just resource stuff and preventing the user from getting fractional resources, can just correct once pr is made with rounding.

**Reply by amylizzle** [reply]
oh yeah, just whack a round() call on absorb and jobs a good 'un. I thought I was missing a super serious issue or something :P

**Comment by amylizzle** [comment]
Just copypasting what I said in discord, please comment if you disagree:
I think the stack thing should just come down to testing out the two and picking the one we like best
now that pali has officially declared fractional amounts to be a bug
and fractional health is basically a non-issue
